enhancements/storage/csi-snapshot.md

```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+
+This feature poses new requirements on Kubernetes distributions such as OpenShift:
+* Ship and install snapshot CRDs.
+* Ship and run snapshot-controller in a cluster.
+  * snapshot-controller is very similar to in-tree PV controller in its function, it watches VolumeSnapshots and provisions VolumeSnapshotContents for them. The snapshot controller is independent on storage backend and only one runs for all CSI drivers installed in a cluster.
+
+Upstream has chosen [addon](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/volumesnapshots) for both. Upstream uses [StatefulSet for the snapshot-controller with single replica](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/volumesnapshots/volume-snapshot-controller/volume-snapshot-controller-deployment.yaml) for the snapshot-controller.
+
 ## Proposal
 
-This feature has already been implemented upstream, and is described in https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190709-csi-snapshot.md . It is currently documented upstream at https://kubernetes.io/docs/concepts/storage/volume-snapshots/ .
+1. OCP ships a new operator for csi-snapshot-controller, say csi-snapshot-controller-operator.
+  * This operator will be installed by CVO in all clusters.
+    * We do not know in advance what CSI drivers will a cluster admin install and we want the cluster ready for snapshots.
+  * This operator will create Deployment with csi-snapshot-controller and report its status via standard `ClusterOperator` object.
+    * Using Deployment with leader election instead of upstream StatefulSet - leader election is significantly faster to run a new leader when a node with the current leader gets unavailable.
+    * TBD: Use DaemonSet on masters? The controller can read / write any VolumeSnapshot, VolumeSnapshotContent and PV objects in the cluster and read any PVC. Running it on masters may be more secure.
+  * Everything runs in namespace `openshift-csi-snapshot-controller`.
+  * Requires new github repo, openshift/csi-snapshot-controller-operator.
+  * Requires new image.
+  * No new CRD / CR is needed for the operator itself, the operator does the same on all clusters / clouds and does not need any configuration.
 
-We will be releasing the `external-snapshotter` sidecar image to enable CSI Snapshot support in OCP as Developer Preview only. This sidecar can be used by internal teams, such as CNV or OCS, for their drivers.
+2. OCP ships new image csi-snapshot-controller. Its source code is already available in github.com/openshift/csi-external-snapshotter, we only need an image.
+  * The component / image is called kubernetes-csi/snapshot-controller upstream, we add csi- prefix to repository and image names, who knows what other kinds of snapshotters there will be.
+
+3. OCP creates upstream snapshot CRDs via CVO.
```

Comment:
if you think of the CVO as an operator of operators, then its responsibility ends with the creation of your operator's CRDs and workload resources.  The CVO shouldn't be used to lay down the resources of your operator's operand.  Such relationships can be hard to transition and your end up heavily constrained by the CVO. See the issues we had with the CVO stripping our required CRD fields earlier this year.

There are existent helpers for "create all these manifests" in library-go and hopefully in controller-runtime.  I recommend using those.

Reply:
This basically means - we shouldn't be packing CRDs used by snapshot-controller inside snapshot-operator's manifests and rather create them when snapshot-operator is started by using libary-go?

This would also mean - when CRDs change, snapshot-operator should be responsible for upgrading them. I think we need to capture this bit.

Reply:
> The CVO shouldn't be used to lay down the resources of your operator's operand.

Ack.